### ISSUE TITLE
As PM, I would like to identify planned features for Release 6.24

### ISSUE THREAD

**Issue by epag** [epag/wresTicketTesting#153]
---

Author Name: **hank.herr** (hank.herr)
Original Redmine Issue: 129950, https://vlab.noaa.gov/redmine/issues\129950
Original Date: 2023-01-10

---

All:

I would like to use this ticket to drive the discussion of features to target for 6.24 delivery; what I've listed below will likely span several releases. I copied it from the 6.22 version. Changes are needed; I'll make those changes next week.

Thanks,

Hank 

===========================================

*+PRIMARY TASKS:+*

* *Move to GitHub*
Top level ticket is #58349. Evan is actively working on this.

* *Wrap Up Recent Developed Features:*
(1) Feature-batched retrieval. (#95867)  Still need to identify appropriate size of batch in production.  50 is too large apparently.

* *Continued Support of Our Users (Miscellany):*
(1) Aliasing variable names. (#130661)
(2) Implement a solution for computing the differences between scores for two modeling scenarios. (#119124)
(3) Add documentation for use of GIS with WRES output (#121272).
(4) Filtering pairs conditionally upon a covariate (i.e., a numeric value that is not one of the pairs, but is expressed at the same valid time/scale): #102096, #102287. 
(5) Event detection (#79741).
(6) Implement forecast progression plot (#45682). 
(7) Implement better HTTPS return codes so that users can know if problems are on their end and if retries are called for.  Possible ticket: #41467.

* *Updates to UI:* React to user requests and changes to the COWRES. 
(1) Overhaul the GUI. (#127599) 
(a) Implement a better approach for the inputs and pairs declaration dialogs (#118789). This might be a byproduct of the above. 
(b) Implement a better feature-specification interface (#73752 is a start).  Should be a table with columns for left/right/baseline; with a dialog that opens to make requests against WRDS to auto-populate the table; the ability to add columns for feature groups with checkboxes to include a feature in the group; etc.
(2) Address registry identified vulnerabilities (#130540) which may be related to EvalStatStore failures (#130711).
(3) Read data from protobuf or another format instead of netCDF to support mapping and the EvalStatStore (#109837).
(4) Other items to address in the "working version" target version.

* *Handling of Features, Integration of the WRDS Location API*:
(1) Fully qualify the WRDS-based thresholds the data model and outputs, thereby properly reacting when a user does not specify a rating curve provider. (#97209)
(2) Implement ability to access WRDS Location API for metadata to fill in gaps in inputs (discussion in #82731).  We may handle this on an as needed basis.

* *Deployment to -prod:*
(1) Apparent Redis bottleneck leading to lost information about evaluation (#122858). Not good.
(2) Scale resources for sampling uncertainty (#120579).
(3) Scale out workers and/or find/fix the bottleneck preventing scale-out (#98818). JPMS/jlink app for scale-out? Graal native-image for scale-out?
(4) Streamline deployment further, automating as much as possible/reasonable.  #88535 as a start.

* *Foundational Improvements* *
(1) Migrate netCDF writing to a microservice. (#129271)
(2) Move external services test to its own, private repository, since its NWC dependent. (#130820)

*+TASKS IN A HOLDING PATTERN PENDING RESOURCES:+*

(1) Support ingest by CHPS (#79998, #86102, #80000).  Implement application to listen to COWRES for evaluation outputs write them to a FEWS friendly format (#86102).  Will use the new messaging API.  Juzer modified a script from Phil in #80000 to use "netcdf2" output, but that change cannot be used at the RFCs until they update their systems to a newer version that allows for a newer version of Python to be used. From NOAT: They are happy with the solution NWRFC provided in #80000, so there is not rush to complete this.  Chris is investigating PI-JSON output from the output service, which could also be a solution.
(2) Prepare additional WRES Cookbook meals, #81204, #79832.  Both are on-hold.  #81204 is an unusual use case and may not merit a meal, because its not using the WRES to conduct an evaluation.  #79832 is waiting on feedback from Taylor, but it might be best to just wait for them to assist us with the pilot performance testing.

+EvalStatStore+
(1) Open the doors to the EvalStatStore to all users (ticket TBD). Temporarily on hold pending additional developer resources.  Currently, the EvalStatStore is hidden by hiding the GUI connections: a *Save Output* button and a checkbox in the scheduling panel.(6) Implement "protobuf" ingest for WRES output service (#88151). netcdf2 is supported, but that only contains scores.  Ingest of protobuf (or csv2) would allow for ingest of diagram data.  Would require extensive testing.

*+PROPOSED USER ACCEPTANCE TESTING:+*

See comments below for details.

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **hank.herr** (hank.herr)
Original Date: 2024-05-23T12:56:37Z

---

Moved planning to 6.24. 

I still need to review the planning items. I hope to do that today, but with a doctor's appointment this afternoon and other work, it may not happen until tomorrow.

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **hank.herr** (hank.herr)
Original Date: 2024-05-30T15:20:29Z

---

Updated the planned tasks. 

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **hank.herr** (hank.herr)
Original Date: 2024-05-30T17:13:02Z

---

Add #130661 to the planning ticket.  

Add #119124, as well. 

For GUI, the top level GUI 2.0 #127599. 

I removed the foundation improvement move to microservices, but netCDF has still not been done (it would be a big on; #129271). We also need to separate out the external services tests to a separate code base/repo. That shouldn't be in GitHub public; maybe private (its NWC specific).  

Also add the ticket for the vulnerabilities the registry is identifying in the GUI.  

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **hank.herr** (hank.herr)
Original Date: 2024-05-31T11:54:40Z

---

Plan updated to reflect current state,

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **hank.herr** (hank.herr)
Original Date: 2024-07-02T18:01:42Z

---

Do we want to perpetuate these planning tickets into GitHub? If not, this is the last such ticket.

The original purpose of these tickets was to provide guidance for developers on high priority tasks that can be picked up when time is available. It was a general roadmap for what's upcoming. However, I'm not sure how often we look at these tickets, so it may not serve that purpose anymore. Plus, there may be better tools for that in GitHub that don't require a ticket. Evan may know. We also have the schedule maintained by Jay that we could make greater use of, though admittedly that schedule has some overhead that may make it more of a pain.  

If you all have any thoughts, please let me know,

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **evan.pagryzinski** (evan.pagryzinski)
Original Date: 2024-07-02T18:03:25Z

---

hank.herr wrote:
> Do we want to perpetuate these planning tickets into GitHub? If not, this is the last such ticket.
> 
> The original purpose of these tickets was to provide guidance for developers on high priority tasks that can be picked up when time is available. It was a general roadmap for what's upcoming. However, I'm not sure how often we look at these tickets, so it may not serve that purpose anymore. Plus, there may be better tools for that in GitHub that don't require a ticket. Evan may know. We also have the schedule maintained by Jay that we could make greater use of, though admittedly that schedule has some overhead that may make it more of a pain.  
> 
> If you all have any thoughts, please let me know,
> 
> Hank

Personally I don't use this ticket super often. With the github project there are built in ticket tracking and planning tools that we could leverage instead
